tests/backup_restore_test.go

```diff
@@ -109,131 +109,169 @@ func TestBackupRestoreBasic(t *testing.T) {
-	CREATE TABLE t1 (
-		id INT,
-		pid INT,
-		PRIMARY KEY (id),
-		UNIQUE INDEX (pid),
-		CONSTRAINT FOREIGN KEY (pid) REFERENCES t0(id) ON DELETE NO ACTION
-	);
-	`)
-	a.NoError(err)
+	validateTbl1 := func(db *sql.DB) {
+		rows, err := db.Query("SELECT * FROM tbl1")
+		a.NoError(err)
+		i := 0
+		for rows.Next() {
+			var col1, col2 int
+			a.NoError(rows.Scan(&col1, &col2))
+			a.Equal(i, col1)
+			a.Equal(i, col2)
+			i++
+		}
+		a.NoError(rows.Err())
+		// TODO(dragonly): change to numRowsTime1 when RestoreIncremental is implemented
+		a.Equal(numRowsTime0, i)
+	}
 
-	// insert data to make time point t0
-	insertData(db, a, 0, 10)
+	// test cases
+	t.Run("Buggy Application", func(t *testing.T) {
+		t.Parallel()
+		ctx := context.Background()
 
-	// make a full backup of t0
-	driverBackup := getDbDriver(a, localhost, fmt.Sprintf("%d", port), username, database)
-	defer func() {
-		err := driverBackup.Close(context.TODO())
-		a.NoError(err)
-	}()
+		t.Log("initialize database")
+		db, stopFn := initDB(t)
+		defer db.Close()
+		defer stopFn()
 
-	buf := backup(a, driverBackup, database)
-	t.Log(buf.String())
+		t.Log("insert data")
+		insertRangeData(t, db, 0, numRowsTime0)
 
-	// insert more data to make time point t1
-	insertData(db, a, 10, 20)
+		t.Log("make a full backup")
+		driver := getMySQLDriver(ctx, t, localhost, fmt.Sprintf("%d", port), username, database)
+		defer func() {
+			err := driver.Close(ctx)
+			a.NoError(err)
+		}()
 
-	// concurrently update data
-	stopChan := make(chan bool)
-	go updateRow(t, username, localhost, database, port, stopChan)
-	defer func() { stopChan <- true }()
+		buf := doBackup(ctx, t, driver, database)
+		t.Logf("backup content:\n%s", buf.String())
 
-	// restore to ghost database
-	_, err = db.Exec("SET foreign_key_checks=OFF")
-	a.NoError(err)
+		t.Log("insert more data")
+		insertRangeData(t, db, numRowsTime0, numRowsTime1)
 
-	_, err = db.Exec(fmt.Sprintf("CREATE DATABASE %s_ghost", database))
-	a.NoError(err)
-	_, err = db.Exec(fmt.Sprintf("USE %s_ghost", database))
-	a.NoError(err)
+		ctxUpdateRow, cancelUpdateRow := context.WithCancel(ctx)
+		t1 := startUpdateRow(ctxUpdateRow, t, username, localhost, database, port)
+		t.Logf("start to concurrently update data at t1: %v", t1)
 
-	driverRestore := getDbDriver(a, localhost, fmt.Sprintf("%d", port), username, fmt.Sprintf("%s_ghost", database))
-	defer func() {
-		err := driverRestore.Close(context.TODO())
+		t.Log("restore to pitr database")
+		timestamp := time.Now().Unix()
+		config := dbPlugin.RecoveryConfig{}
+		err := driver.RestorePITR(ctx, bufio.NewScanner(buf), config, database, timestamp)
 		a.NoError(err)
-	}()
-	err = driverRestore.Restore(context.TODO(), bufio.NewScanner(buf))
-	a.NoError(err)
-
-	// TODO(dragonly): validate ghost table data and schema
 
-	// TODO(dragonly): apply binlog from full backup to
-	// need to use binlog package from gh-ost or go-mysql
+		t.Log("cutover stage")
+		cancelUpdateRow()
+		time.Sleep(time.Second)
```

Comment:
Any reason we need to sleep for a second? Otherwise, we need to comment this.

Reply:
To wait for the ongoing update process to shut down. This mimics the operation of a user manual cutover, where the user must wait for the database idle before doing the cutover.
Added comments.